app/services/analysis/gwas/genotype_hapmap_parser.rb

```diff
@@ -0,0 +1,87 @@
+class Analysis
+  class Gwas
+    class GenotypeHapmapParser
+      HEADER_COLUMNS = %w(rs# alleles chrom pos strand assembly# center protLSID assayLSID panelLSID QCcode)
+
+      def call(io)
+        Result.new(io).tap do |result|
+          if result.columns[0...HEADER_COLUMNS.size] != HEADER_COLUMNS
+            result.errors << :not_a_hapmap
+          end
+
+          if result.sample_ids.blank?
+            result.errors << :no_samples
+          end
+
+          if io.eof?
+            result.errors << :no_mutations
+          end
+        end
+      end
+
+      private
+
+      class Result
+        attr_accessor :columns
+        attr_reader :errors, :io
+
+        def initialize(io)
+          @io = io
+          @errors = []
+          @columns = io.readline.split(/\s+/)
+        rescue EOFError
+          @columns = []
+        end
+
+        def valid?
+          errors.empty?
+        end
+
+        def sample_ids
+          @sample_ids ||= columns[HEADER_COLUMNS.size..-1]
+        end
+
+        def each_record(&blk)
+          enum = io.each_line.lazy.map { |line| Record.new(sample_ids, line.split(/\s+/)) }
```

Comment:
Can we have a more descriptive name instead of `enum`?

Reply:
Changed to `records`.